streaming_data_types/area_detector_NDAr.py

```diff
@@ -0,0 +1,75 @@
+from typing import Union
+import flatbuffers
+from streaming_data_types.fbschemas.NDAr_NDArray_schema import NDArray
+from streaming_data_types.utils import check_schema_identifier
+from collections import namedtuple
+import time
+
+FILE_IDENTIFIER = b"NDAr"
+
+
+def serialise_ndar(
+    id: str,
+    dims: list,
+    data_type: int,
+    data: list,
+) -> bytes:
+    builder = flatbuffers.Builder(1024)
+    builder.ForceDefaults(True)
+
+    # Build dims
+    NDArray.NDArrayStartDimsVector(builder, len(dims))
+    # FlatBuffers builds arrays backwards
+    for s in reversed(dims):
+        builder.PrependUint64(s)
+    dims_offset = builder.EndVector(len(dims))
+
+    # Build data
+    NDArray.NDArrayStartPDataVector(builder, len(data))
+    # FlatBuffers builds arrays backwards
+    for s in reversed(data):
+        builder.PrependUint8(s)
+    data_offset = builder.EndVector(len(data))
+
+    # Build the actual buffer
+    NDArray.NDArrayStart(builder)
+    NDArray.NDArrayAddDataType(builder, data_type)
+    NDArray.NDArrayAddDims(builder, dims_offset)
+    NDArray.NDArrayAddId(builder, id)
+    NDArray.NDArrayAddPData(builder, data_offset)
+    NDArray.NDArrayAddTimeStamp(builder, int(time.time() * 1000))
+    nd_array_message = NDArray.NDArrayEnd(builder)
+
+    builder.Finish(nd_array_message, file_identifier=FILE_IDENTIFIER)
+    return bytes(builder.Output())
+
+
+nd_Array = namedtuple(
+    "NDArray",
+    (
+        "data_type",
+        "id",
+        "timestamp",
+        "dims",
+        "data",
+    ),
+)
+
+
+def deserialise_ndar(buffer: Union[bytearray, bytes]) -> NDArray:
+    check_schema_identifier(buffer, FILE_IDENTIFIER)
+
+    nd_array = NDArray.NDArray.GetRootAsNDArray(buffer, 0)
+    id = nd_array.Id() if nd_array.Id() else b""
+    timestamp = nd_array.TimeStamp() if nd_array.TimeStamp() else b""
+    data_type = nd_array.DataType() if nd_array.DataType() else b""
+    dims = nd_array.DimsAsNumpy()
+    data = nd_array.PDataAsNumpy()
```

Comment:
It would probably be nicer to the end user to convert the data (numpy) array into the type that is indicated by the `data_type` variable (example for how to do this at the end of the comment). However, seeing as you are the only user for now, I am ok with letting you implement that part elsewhere.

```
def GetDataArr(self, fb_arr):
  nr_of_bytes = fb_arr.PDataLength()
  raw_data = fb_arr.PDataAsNumpy()
  dim_list = [fb_arr.Dims(0), ]
  for i in range(1, fb_arr.DimsLength()):
    dim_list.append(fb_arr.Dims(i))
  numpy_arr_type = [np.int8, np.uint8, np.int16, np.uint16, np.int32, np.uint32, np.float32, np.float64]
  data_arr = raw_data.view(numpy_arr_type[fb_arr.DataType()]).reshape(dim_list)
  return data_arr
```

Reply:
This is much more elegant than the way I was doing it, thanks